src/components/10-atoms/input-text/index.js

```diff
@@ -54,14 +65,72 @@ class AXAInputText extends AXAPopupMixin(NoShadowDOM) {
     this.invalid = false;
     this.disabled = false;
     this.isReact = false;
+    this.modelCounter = '';
     this.onFocus = () => {};
     this.onBlur = () => {};
     this.onChange = () => {};
 
+    this.counter = '';
+    this.counterMax = '';
+
     // internal properties
     this.nativeInput = { value: '' };
     this.modelValue = '';
     this.isControlled = false;
+    this.isPlaceholderInCounter = false;
+  }
+
+  get charsLeft() {
+    const {
+      maxLength,
+      nativeInput: { value: nativeValue },
+    } = this;
+
+    if (nativeValue) {
+      return maxLength - nativeValue.length;
+    }
+
+    return maxLength;
+  }
+
+  get areCharsLeft() {
+    return this.charsLeft > 0;
+  }
+
+  get showCounter() {
+    return (
+      this.maxLength > 0 && !this.invalid && this.areCharsLeft && !this.disabled
+    );
+  }
+
+  get showCounterMax() {
+    return (
+      this.maxLength > 0 &&
+      this.counterMax &&
+      !this.showError &&
+      !this.areCharsLeft &&
+      !this.disabled
+    );
+  }
+
+  get showMessages() {
+    return this.showError || this.showCounter || this.showCounterMax;
+  }
+
+  get replaceCounterPlaceholder() {
+    return this.counter.replace(/##.*##/, this.charsLeft - 1);
```

Comment:
Keep it DRY: I suggest `const userCharsLeft = this.charsLeft - 1` and then use it everywhere where you'd subtract 1 now.

Reply:
To implement your suggestion, I got rid of one helper method and inlined it. So I can make that a local variable instead of a class level variable, because that would have been confusing in my opinion. Having something like `charsLeft` and `charsLeftUntilMaxLength` on the class level would make a lot of developers scratch their heads.